resource-aggregate/cqrs/eventstore/cqldb/load.go

```diff
@@ -0,0 +1,228 @@
+package cqldb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gocql/gocql"
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+	"github.com/plgd-dev/hub/v2/pkg/cqldb"
+	pkgTime "github.com/plgd-dev/hub/v2/pkg/time"
+	"github.com/plgd-dev/hub/v2/resource-aggregate/cqrs/eventstore"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type iterator struct {
+	iter            *gocql.Iter
+	dataUnmarshaler UnmarshalerFunc
+	err             error
+}
+
+const errFmtDataIsNotStringType = "invalid data['%v'] type ('%T'), expected string type"
+
+func newIterator(iter *gocql.Iter, dataUnmarshaler UnmarshalerFunc) *iterator {
+	return &iterator{
+		iter:            iter,
+		dataUnmarshaler: dataUnmarshaler,
+	}
+}
+
+func (i *iterator) nextEvent(_ context.Context) map[string]interface{} {
+	if i.err != nil {
+		return nil
+	}
+	v := make(map[string]interface{})
+	ok := i.iter.MapScan(v)
+	if !ok {
+		return nil
+	}
+	return v
+}
+
+func (i *iterator) Next(ctx context.Context) (eventstore.EventUnmarshaler, bool) {
+	ev := i.nextEvent(ctx)
+	if ev == nil {
+		return nil, false
+	}
+	eventType, ok := ev[eventTypeKey].(string)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event is not a string but %T", eventTypeKey, ev[eventTypeKey])
+		return nil, false
+	}
+	isSnapshot := true
+	timestamp, ok := ev[timestampKey].(int64)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event %v is not an int64 but %T", timestampKey, eventTypeKey, ev[timestampKey])
+		return nil, false
+	}
+	version, ok := ev[versionKey].(int64)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event %v is not an int64 but %T", versionKey, eventTypeKey, ev[versionKey])
+		return nil, false
+	}
+	aggregateID, ok := ev[idKey].(gocql.UUID)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event %v is not a string but %T", idKey, eventTypeKey, ev[idKey])
+		return nil, false
+	}
+	groupID, ok := ev[deviceIDKey].(gocql.UUID)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event %v is not a string but %T", deviceIDKey, eventTypeKey, ev[deviceIDKey])
+		return nil, false
+	}
+	snapshot, ok := ev[snapshotKey].([]byte)
+	if !ok {
+		i.err = fmt.Errorf("invalid data, '%v' of event %v is not a []byte but %T", snapshotKey, eventTypeKey, ev[snapshotKey])
+		return nil, false
+	}
+	return eventstore.NewLoadedEvent(
+		uint64(version),
```

Comment:
why the first conversion to `int64` on line 61 and then this second cast? can't it be `uint64` from the start?

Reply:
No because, cqldb doesn't support `uint64`